# K-Nearest Neighbours Classification of Heart Disease

Data downloaded from Kaggle at https://www.kaggle.com/ronitf/heart-disease-uci

## Importing Packages

In [1]:
##Import packages

#Data Processing
import pandas as pd
import numpy as np

#Finding the data on our system
import pathlib

#KNN
from sklearn.neighbors import KNeighborsClassifier

#Scaling
from sklearn.preprocessing import StandardScaler

#Splitting data into training and test
from sklearn.model_selection import train_test_split

#Metrics to evaluate
from sklearn.metrics import accuracy_score

#Plotting library
import matplotlib.pyplot as plt

## Import Data

In [2]:
##Import data from another directory

#Find the file from a local directiory
p = str(pathlib.Path().resolve()).replace("\\Notebooks\\Classification Methods", "")
p += ("\\Data\\Raw\\Heart_Disease_UCI.csv")
p = pathlib.Path(p).resolve()

#Read into a pandas dataframe
df = pd.read_csv(p)

#Preview the dataset
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

### Understanding the Columns in the Data

When you collect your own data, you should know exactly what each of the variables represent, and how the variables are encoded. Unfortunately, when working with others' data, we don't start with a full understanding of the data. Not being able to understand the data makes it extremely difficult to gain insight from it. You should definitely document any data of your own, as you might have to come back to it later, or someone else might have to interpret it in the future.

Luckily for this dataset, there is an explanation of each column, that we should look at to gain an insight into what each variable represents. There are some missing pieces though, and we might want to think about dropping these columns, if we can't reasonably collect any more data with the same variables.
- age: Age in years
- sex: (1 = male; 0 = female)
- cp: chest pain type
- trestbps: resting blood pressure (in mm Hg on admission to the hospital)
- chol: serum cholestoral in mg/dl
- fbs: (fasting blood sugar &gt; 120 mg/dl) (1 = true; 0 = false)
- restecg: resting electrocardiographic results
- thalach: maximum heart rate achieved
- exang: exercise induced angina (1 = yes; 0 = no)
- oldpeak: ST depression induced by exercise relative to rest
- slope: the slope of the peak exercise ST segment
- ca: number of major vessels (0-3) colored by flourosopy
- thal: 3 = normal; 6 = fixed defect; 7 = reversable defect
- target: 1 or 0 (Dependent, does the patient have a heart defect - 0 = Disease, 1 = No Disease)


We can also view this information on Kaggle, where histograms of the distribution of every variable are visible, along with basic summary statistics. This is a good starting point for exploratory data analysis! Be sure to understand the data that you're working with.

However, we need to be careful with this data - by inspecting the source it is found that the full description of the data isn't accurate! This is best detailed in [this discussion post](https://www.kaggle.com/ronitf/heart-disease-uci/discussion/105877): https://www.kaggle.com/ronitf/heart-disease-uci/discussion/105877. Here we have another illustration of why exploring datasets that don't belong to you is so important.